### PR TITLE
feat!: overhaul slurmrestd charm API

### PR DESCRIPTION
**Pull request by jamesbeedy**
Please reference the [Slurm Charms Change Summary Document](https://docs.google.com/document/d/1bP95EXfM-1qp81IF1CA4igLVZIKnsjWdHdN6PZpEh74/edit) for an in-depth explanation of these changes.

Changes include:
* remove jinja2 and replace with `Path().write_text()`
* remove `slurm-ops-manager` 3rd party dep and replace with local `slurmrestd_ops.py`
* consolidate the yaml files into a single `charmcraft.yaml`
* rename interface slurmrestd -> slurmctld
* cleanup from PR review
* remove unused code
* addeded logging
* simplified api
* reduced code footprint
* consolidated yaml files into charmcraft.yaml
* added type checking

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,11 +30,12 @@ $ juju deploy slurmd --channel edge
 $ juju deploy slurmdbd --channel edge
 $ juju deploy mysql --channel 8.0/edge
 $ juju deploy mysql-router slurmdbd-mysql-router --channel dpe/edge
-$ juju integrate slurmctld:slurmrestd slurmrestd:slurmrestd
-$ juju integrate slurmctld:slurmd slurmd:slurmd
+
```
Comment (NucciTheBoss) [inline README.md:33]: [nit]: I like to keep code blocks together :smile: 

```suggestion
```

File: src/slurmrestd_ops.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        """Return the slurm repo."""
+        ppa_url: str = "https://ppa.launchpadcontent.net/ubuntu-hpc/slurm-wlm-23.02/ubuntu"
+        sources_list: str = (
+            f"deb [signed-by={self._keyring_path}] {ppa_url} {distro.codename()} main"
+        )
+        return apt.DebianRepository.from_repo_line(sources_list)
+
+    def install(self) -> bool:
+        """Install the slurmrestd package using lib apt."""
+        # Install the key.
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:80]: [nit]: I don't think this comment is necessary

```suggestion
```

File: src/slurmrestd_ops.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        if self._keyring_path.exists():
+            self._keyring_path.unlink()
+        self._keyring_path.write_text(SLURM_PPA_KEY)
+
+        # Add the repo.
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:85]:
```suggestion
```

File: src/templates/slurm.conf.tmpl
Comment (NucciTheBoss) [inline src/templates/slurm.conf.tmpl:1]: Is this file necessary to ship with the charm? I don't see it referenced anywhere in the charm code. I also believe that we've dropped `jinja2` as dependency :eyes:

File: src/slurmrestd_ops.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+    @property
+    def _slurmrestd_group_id(self) -> str:
+        """Return the slurmrestd group ID."""
+        return "64031"
+
+    @property
+    def _slurmrestd_group_name(self) -> str:
+        """Return the slurmrestd group."""
+        return "slurmrestd"
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:333]: Would it possibly be better to have these set as constants within a `constants.py` module rather than have them be object properties? Since this isn't a public package, I'm not worried about folks potentially overwriting the values.

File: src/slurmrestd_ops.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+            status_str = status.decode().strip()
+            if "active" in status_str:
+                logger.debug("#### Munge daemon active")
+                return True
+            else:
+                logger.error(f"## Munge not running: {status_str}")
+                return False
+        except subprocess.CalledProcessError as e:
+            logger.error(f"## Error querring munged - {e}")
+            return False
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:203]: Made a similar comment on the slurmd operator IIRC, but I don't think we need all this logic here as it can be superseded by the `service_running(...)` function in the `systemd` charm library:

```suggestion
        return systemd.service_running("munged")
```

In fact, you might not even need this method as you can just use the `service_running(...)` function to get the current state of the `munged` service.

File: src/slurmrestd_ops.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+            elif action == "stop":
+                systemd.service_stop(service)
+            elif action == "restart":
+                systemd.service_restart(service)
+            else:
+                logger.debug("Only start, stop, restart supported.")
+        except SlurmrestdException("Cannot restart slurmrestd.") as e:  # type: ignore [misc]
+            logger.error(e)
+            return False
+        return True
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:254]: Hmm... not sure that we should have a direct interface to systemd like this especially since it seems like a match-case block for various actions that we can perform against the `slurmrestd` service. I also noticed that we're using the same method for managing `munged` service even though the docstring indicates that this method is only for managing the `slurmrestd` service.

I think it would be better if we had separate methods for starting, stopping, and restarting - `start()`, `stop()`, `restart()` respectively - the `slurmrestd` and `munged` service. That way we can tightly control the input and don't have to worry about accidentally passing an incorrect action or service name to the method.

File: src/slurmrestd_ops.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        """Uninstall the slurmrestd package using libapt."""
+        if apt.remove_package(self._package_name):
+            logger.info(f"{self._package_name} removed from system.")
+        else:
+            logger.error(f"{self._package_name} not found on system")
+
+        repositories = apt.RepositoryMapping()
+        repositories.disable(self._repo())
+
+        # Remove the key.
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:111]:
```suggestion
```

File: src/slurmrestd_ops.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        systemd.service_enable(self._munged_systemd_service)
+        systemd.service_enable(self._slurmrestd_systemd_service)
+
+        systemd.service_stop(self._munged_systemd_service)
+
+        if not self._slurm_conf_dir.exists():
+            self._slurm_conf_dir.mkdir()
+
+        user = f"{self._slurmrestd_user_name}:{self._slurmrestd_group_name}"
+        subprocess.call(["chown", "-R", user, self._slurm_conf_dir])
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:148]: Any reason to not use `os.chown(...)` here? Since you're creating the slurmrestd user and group within the charm, it wouldn't be too difficult to retrieve the uid and gid of the slurmrestd user using the `pwd` and `grp` modules :smiley:

File: requirements.txt [hunk 9]
```diff
@@ -1,2 +1,2 @@
-ops==2.*
-git+https://github.com/omnivector-solutions/slurm-ops-manager.git@0.8.16
\ No newline at end of file
+ops
+distro
```
Comment (NucciTheBoss) [inline requirements.txt:2]: Should pin the dependency versions here

File: src/slurmrestd_ops.py [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+from base64 import b64decode
+from pathlib import Path
+
+import charms.operator_libs_linux.v0.apt as apt  # type: ignore [import-untyped]
+import charms.operator_libs_linux.v1.systemd as systemd  # type: ignore [import-untyped]
+import distro
+
+logger = logging.getLogger()
+
+
```
Comment (NucciTheBoss) [inline src/slurmrestd_ops.py:18]: You can have logger and the local constant together here

```suggestion
```

**Review by NucciTheBoss — CHANGES_REQUESTED** [review]
Overall, looks great to me! Just a couple of comments and nits

**Comment by NucciTheBoss** [comment]
Hey @jamesbeedy :wave: 

Similar to the slurmd operator, could you please apply the changes you made to the slurmdbd and slurmctld operators here as well before merging? We can focus on feature refinement in follow-on pull requests :smile:

**Comment by jamesbeedy** [comment]
@NucciTheBoss g2g on my end

File: src/slurmrestd_ops.py (hunk 10 above)
Comment (jamesbeedy) [reply src/slurmrestd_ops.py:26]: Just going to leave as black likes it for now. We can change it later to your liking ... I don't mind.

**Review by NucciTheBoss — APPROVED** [review]
LGTM! Thanks for your work @jamesbeedy :tada:
